File: backend/src/main/java/com/joying/chat/service/ChatMessageService.java
package com.joying.chat.service;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;
import java.util.Set;
import java.util.Objects;
import java.util.Map;
import java.util.HashMap;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

import com.joying.chat.document.ChatMessage;
import com.joying.chat.document.MessageType;
import com.joying.chat.domain.ChatRoom;
import com.joying.chat.dto.ChatMessageResponse;
import com.joying.chat.repository.ChatMessageRepository;
import com.joying.chat.repository.ChatRoomRepository;
import com.joying.common.exception.BusinessException;
import com.joying.common.exception.ErrorCode;

import lombok.RequiredArgsConstructor;

/**
 * 메시지를 읽고 고치고 지운다.
 *
 * <p>보내는 것은 {@link ChatService}가 맡는다. 여기는 이미 저장된 것을 다룬다.
 */
@Service
@RequiredArgsConstructor
public class ChatMessageService {

	private static final Logger log = LoggerFactory.getLogger(ChatMessageService.class);

	private static final String SEQUENCE = "sequence";

	private final ChatMessageRepository chatMessageRepository;
	private final ChatRoomRepository chatRoomRepository;
	private final RedisPubSubPublisher redisPubSubPublisher;
	private final ChatRoomPermissionCache permissionCache;

	/**
	 * 이 방을 볼 수 있는 사람인지.
	 *
	 * <p>보내는 경로와 같은 판정을 쓴다. 예전에는 여기서 구매자와 판매자인지만 보고
	 * 나갔는지를 보지 않아, 나간 사람이 히스토리를 계속 읽을 수 있었다.
	 */
	private void validateChatRoomAccess(Long chatRoomId, Long memberId) {
		if (!permissionCache.hasPermission(chatRoomId, memberId)) {
			throw new BusinessException(ErrorCode.FORBIDDEN, "채팅방 접근 권한이 없습니다");
		}
	}

	public List<ChatMessageResponse> getMessages(Long chatRoomId, int page, int size) {
		return toResponses(chatMessageRepository
			.findByChatRoomIdAndIsDeletedFalseOrderBySequenceDesc(chatRoomId, desc(page, size)));
	}

	/**
	 * 커서로 메시지를 가져온다.
	 *
	 * <p>{@code before}는 과거로 거슬러 올라가는 것이고 {@code after}는 끊긴 사이에
	 * 놓친 것을 받는 것이다. 방향이 반대라 함께 쓸 수 없다. 정렬도 그래서 다르다.
	 * 놓친 것은 받는 쪽이 순서대로 이어붙여야 하므로 오래된 순으로 준다.
	 *
	 * <p>커서는 방 안의 메시지 번호다. 시각을 커서로 쓰면 같은 밀리초에 저장된 것이
	 * 경계에서 빠지거나 두 번 온다.
	 */
	public List<ChatMessageResponse> getMessagesBefore(Long chatRoomId, Long before,
													   Long after, int size, Long memberId) {
		validateChatRoomAccess(chatRoomId, memberId);

		if (before != null && after != null) {
			throw new BusinessException(ErrorCode.INVALID_INPUT_VALUE,
				"before와 after 파라미터는 동시에 사용할 수 없습니다");
		}

		List<ChatMessage> messages;
		if (after != null) {
			messages = chatMessageRepository
				.findByChatRoomIdAndIsDeletedFalseAndSequenceGreaterThanOrderBySequenceAsc(
					chatRoomId, after, asc(0, size));
		} else if (before != null) {
			messages = chatMessageRepository
				.findByChatRoomIdAndIsDeletedFalseAndSequenceLessThanOrderBySequenceDesc(
					chatRoomId, before, desc(0, size));
		} else {
			messages = chatMessageRepository
				.findByChatRoomIdAndIsDeletedFalseOrderBySequenceDesc(chatRoomId, desc(0, size));
		}

		return toResponses(messages);
	}

	public List<ChatMessageResponse> searchMessages(Long chatRoomId, String keyword,
													int page, int size, Long memberId) {
		validateChatRoomAccess(chatRoomId, memberId);

		return toResponses(chatMessageRepository
			.findByChatRoomIdAndIsDeletedFalseAndContentContainingOrderBySequenceDesc(
				chatRoomId, keyword, desc(page, size)));
	}

	/**
	 * 안읽음 건수.
	 *
	 * <p>읽은 시각을 받아 그 이후를 센다. 정확히 세는 것은 {@code UnreadCountService}
	 * 가 읽은 번호로 하고, 여기는 시각만 아는 자리에서 쓴다.
	 */
	public long getUnreadCount(Long chatRoomId, Instant lastReadAt) {
		if (lastReadAt == null) {
			return 0L;
		}
		return chatMessageRepository
			.countByChatRoomIdAndIsDeletedFalseAndCreatedAtAfter(chatRoomId, lastReadAt);
	}

	/**
	 * 끊긴 사이에 놓친 것을 받는다.
	 */
	public List<ChatMessageResponse> getMessagesAfter(Long chatRoomId, Long after,
													  int limit, Long memberId) {
		validateChatRoomAccess(chatRoomId, memberId);

		return toResponses(chatMessageRepository
			.findByChatRoomIdAndIsDeletedFalseAndSequenceGreaterThanOrderBySequenceAsc(
				chatRoomId, after, asc(0, limit)));
	}

	/**
	 * 특정 메시지를 가운데 두고 앞뒤를 함께 준다.
	 *
	 * <p>답장을 눌러 원본으로 뛰거나 검색 결과에서 문맥을 볼 때 쓴다. 앞엣것은 최신순으로
	 * 받아 뒤집어야 시간순이 된다.
	 */
	public List<ChatMessageResponse> getMessagesAround(Long chatRoomId, String messageId,
													   int before, int after, Long memberId) {
		validateChatRoomAccess(chatRoomId, memberId);

		ChatMessage target = chatMessageRepository.findById(messageId)
			.orElseThrow(() -> new BusinessException(
				ErrorCode.RESOURCE_NOT_FOUND, "메시지를 찾을 수 없습니다"));

		if (!chatRoomId.equals(target.getChatRoomId())) {
			throw new BusinessException(ErrorCode.INVALID_INPUT_VALUE, "해당 채팅방의 메시지가 아닙니다");
		}

		Long targetSequence = target.getSequence();
		if (targetSequence == null) {
			throw new BusinessException(ErrorCode.INTERNAL_SERVER_ERROR, "메시지 번호가 없습니다");
		}

		List<ChatMessage> all = new ArrayList<>();

		if (before > 0) {
			List<ChatMessage> beforeMessages = new ArrayList<>(chatMessageRepository
				.findByChatRoomIdAndIsDeletedFalseAndSequenceLessThanOrderBySequenceDesc(
					chatRoomId, targetSequence, desc(0, before)));
			// 최신순으로 받았으므로 시간순으로 뒤집는다
			java.util.Collections.reverse(beforeMessages);
			all.addAll(beforeMessages);
		}

		all.add(target);

		if (after > 0) {
			all.addAll(chatMessageRepository
				.findByChatRoomIdAndIsDeletedFalseAndSequenceGreaterThanOrderBySequenceAsc(
					chatRoomId, targetSequence, asc(0, after)));
		}

		return toResponses(all);
	}

	/**
	 * 지운 것으로 표시한다.
	 *
	 * <p>실제로 지우지 않는 이유는 답장이 이 메시지를 가리키고 있을 수 있기 때문이다.
	 * 지우면 상대 화면의 답장이 무엇을 가리키는지 알 수 없게 된다.
	 */
	public void deleteMessage(Long chatRoomId, String messageId, Long memberId) {
		ChatMessage message = requireOwnMessage(chatRoomId, messageId, memberId);

		if (message.isDeleted()) {
			throw new BusinessException(ErrorCode.INVALID_INPUT_VALUE, "이미 삭제된 메시지입니다");
		}

		message.delete();
		ChatMessage saved = chatMessageRepository.save(message);

		log.info("메시지 삭제 완료: messageId={}, chatRoomId={}, memberId={}",
			messageId, chatRoomId, memberId);

		publishToBothSides(chatRoomId, memberId, saved);
	}

	public ChatMessageResponse updateMessage(Long chatRoomId, String messageId,
											 Long memberId, String newContent) {
		if (newContent == null || newContent.isBlank()) {
			throw new BusinessException(ErrorCode.INVALID_INPUT_VALUE, "메시지 내용은 비어 있을 수 없습니다");
		}

		ChatMessage message = requireOwnMessage(chatRoomId, messageId, memberId);

		if (message.getType() != MessageType.TEXT) {
			throw new BusinessException(ErrorCode.INVALID_INPUT_VALUE, "텍스트 메시지만 수정할 수 있습니다");
		}
		if (message.isDeleted()) {
			throw new BusinessException(ErrorCode.INVALID_INPUT_VALUE, "삭제된 메시지는 수정할 수 없습니다");
		}
		if (newContent.equals(message.getContent())) {
			throw new BusinessException(ErrorCode.INVALID_INPUT_VALUE, "수정할 내용이 기존 내용과 동일합니다");
		}

		boolean firstEdit = message.edit(newContent, Instant.now());
		ChatMessage saved = chatMessageRepository.save(message);

		log.info("메시지 수정 완료: messageId={}, chatRoomId={}, memberId={}, 첫 수정={}",
			messageId, chatRoomId, memberId, firstEdit);

		return publishToBothSides(chatRoomId, memberId, saved);
	}

	/**
	 * 이 사람이 보낸 메시지가 맞는지 확인하고 돌려준다.
	 *
	 * <p>고치기와 지우기가 같은 확인을 하고 있어 한곳으로 모았다.
	 */
	private ChatMessage requireOwnMessage(Long chatRoomId, String messageId, Long memberId) {
		ChatMessage message = chatMessageRepository.findById(messageId)
			.orElseThrow(() -> new BusinessException(
				ErrorCode.RESOURCE_NOT_FOUND, "메시지를 찾을 수 없습니다"));

		if (!chatRoomId.equals(message.getChatRoomId())) {
			throw new BusinessException(ErrorCode.INVALID_INPUT_VALUE, "해당 채팅방의 메시지가 아닙니다");
		}
		if (!memberId.equals(message.getSenderId())) {
			throw new BusinessException(ErrorCode.FORBIDDEN, "본인의 메시지만 조작할 수 있습니다");
		}
		return message;
	}

	/**
	 * 바뀐 메시지를 양쪽 화면에 반영한다.
	 *
	 * <p>받는 사람을 미리 실어 보내면 받는 서버가 방을 다시 조회하지 않아도 된다.
	 */
	private ChatMessageResponse publishToBothSides(Long chatRoomId, Long memberId,
												   ChatMessage saved) {
		ChatMessage replyMessage = findReplyOf(saved);

		ChatRoom chatRoom = chatRoomRepository.findById(chatRoomId)
			.orElseThrow(() -> new BusinessException(
				ErrorCode.RESOURCE_NOT_FOUND, "채팅방을 찾을 수 없습니다"));

		Long receiverId = memberId.equals(chatRoom.getBuyer().getMemberId())
			? chatRoom.getSeller().getMemberId()
			: chatRoom.getBuyer().getMemberId();

		ChatMessageResponse response = ChatMessageResponse.from(saved, replyMessage)
			.toBuilder()
			.receiverId(receiverId)
			.build();

		redisPubSubPublisher.publish(response);
		return response;
	}

	/**
	 * 답장 대상 메시지. 답장이 아니면 null.
	 *
	 * <p>한 건을 내보낼 때만 쓴다. 목록에는 쓰지 않는다.
	 */
	private ChatMessage findReplyOf(ChatMessage message) {
		if (message.getReplyToMessageId() == null) {
			return null;
		}
		return chatMessageRepository.findById(message.getReplyToMessageId()).orElse(null);
	}

	/**
	 * 목록에 실을 답장 대상을 한 번에 가져온다.
	 *
	 * <p>예전에는 메시지마다 따로 조회했다. 답장이 섞여 있으면 한 번 읽을 때 그만큼
	 * 조회가 더 나간다. 500건을 읽을 때 답장이 없으면 p95 10ms, 전부 답장이면 60ms 였다.
	 *
	 * <p>같은 메시지에 여러 건이 답장할 수 있어 식별자를 모아 중복을 없앤 뒤 부른다.
	 * 지워진 것을 가리키는 답장은 결과에 없으므로 {@code null} 이 되고, 화면은 그것을
	 * 지워진 메시지로 표시한다.
	 */
	private Map<String, ChatMessage> replyTargetsOf(List<ChatMessage> messages) {
		Set<String> ids = messages.stream()
			.map(ChatMessage::getReplyToMessageId)
			.filter(Objects::nonNull)
			.collect(Collectors.toSet());

		if (ids.isEmpty()) {
			return Map.of();
		}

		Map<String, ChatMessage> byId = new HashMap<>();
		chatMessageRepository.findAllById(ids).forEach(message -> byId.put(message.getId(), message));
		return byId;
	}

	private List<ChatMessageResponse> toResponses(List<ChatMessage> messages) {
		Map<String, ChatMessage> replyTargets = replyTargetsOf(messages);

		return messages.stream()
			.map(message -> {
				String replyToId = message.getReplyToMessageId();
				ChatMessage replyTarget = replyToId == null ? null : replyTargets.get(replyToId);
				return ChatMessageResponse.from(message, replyTarget);
			})
			.toList();
	}

	private Pageable desc(int page, int size) {
		return PageRequest.of(page, size, Sort.by(Sort.Direction.DESC, SEQUENCE));
	}

	private Pageable asc(int page, int size) {
		return PageRequest.of(page, size, Sort.by(Sort.Direction.ASC, SEQUENCE));
	}
}

File: backend/src/main/kotlin/com/joying/chat/service/ChatMessageService.kt
package com.joying.chat.service

import com.joying.chat.document.MessageType
import com.joying.chat.dto.ChatMessageResponse
import com.joying.chat.repository.ChatMessageRepository
import com.joying.chat.repository.ChatRoomRepository
import com.joying.common.exception.BusinessException
import com.joying.common.exception.ErrorCode
import org.slf4j.LoggerFactory
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Sort
import org.springframework.stereotype.Service
import java.time.Instant

/**
 * 채팅 메시지 Service
 *
 * 저장소가 전부 블로킹이라 그대로 부른다. 서로 의존하지 않는 조회만 동시에 날린다.
 */
@Service
class ChatMessageService(
    private val chatMessageRepository: ChatMessageRepository,
    private val chatRoomRepository: ChatRoomRepository,
    private val redisPubSubPublisher: RedisPubSubPublisher
) {
    private val logger = LoggerFactory.getLogger(ChatMessageService::class.java)

    /**
     * 채팅방 접근 권한 확인
     *
     * @param chatRoomId 채팅방 ID
     * @param memberId 회원 ID
     * @throws BusinessException 권한이 없는 경우
     */
    private fun validateChatRoomAccess(chatRoomId: Long, memberId: Long) {
        val chatRoom = chatRoomRepository.findById(chatRoomId)
            .orElseThrow { BusinessException(ErrorCode.RESOURCE_NOT_FOUND, "채팅방을 찾을 수 없습니다") }

        // 구매자 또는 판매자만 접근 가능
        if (chatRoom.buyer.getMemberId() != memberId && chatRoom.seller.getMemberId() != memberId) {
            throw BusinessException(ErrorCode.FORBIDDEN, "채팅방 접근 권한이 없습니다")
        }
    }

    /**
     * 채팅방의 메시지 목록 조회 (페이징)
     *
     * @param chatRoomId 채팅방 ID
     * @param page 페이지 번호 (0부터 시작)
     * @param size 페이지 크기
     * @return 메시지 목록 (최신순)
     */
    fun getMessages(chatRoomId: Long, page: Int, size: Int): List<ChatMessageResponse> =
        run {
            val pageable = PageRequest.of(page, size, Sort.by(Sort.Direction.DESC, "createdAt"))

            chatMessageRepository
                .findByChatRoomIdAndIsDeletedFalseOrderByCreatedAtDesc(chatRoomId, pageable)
                .map { message ->
                    val replyMessage = message.replyToMessageId?.let { replyId ->
                        chatMessageRepository.findById(replyId).orElse(null)
                    }
                    ChatMessageResponse.from(message, replyMessage)
                }
        }

    /**
     * 커서 기반 페이징으로 메시지 조회 (통합)
     *
     * - before 파라미터: 과거 메시지 조회 (무한 스크롤)
     * - after 파라미터: 놓친 메시지 조회 (재연결 시)
     *
     * 예전에는 이 자리를 코루틴이 맡았다. 밑에 있는 저장소가 전부 블로킹이라
     * 코루틴이 사 주는 것이 없었고, 톰캣 워커를 잡아 둔 채 IO 스레드를 하나 더
     * 쓰고 있었다. 지금은 그냥 부른다.
     *
     * @param chatRoomId 채팅방 ID
     * @param before 이 시간 이전의 메시지 조회 (과거 방향, 최신순 정렬)
     * @param after 이 시간 이후의 메시지 조회 (놓친 메시지, 오래된 순 정렬)
     * @param size 가져올 개수
     * @param memberId 요청한 회원 ID (권한 확인용)
     * @return 메시지 목록
     */
    fun getMessagesBefore(
        chatRoomId: Long,
        before: Instant? = null,
        after: Instant? = null,
        size: Int,
        memberId: Long
    ): List<ChatMessageResponse> = run {
        // 권한 확인
        validateChatRoomAccess(chatRoomId, memberId)

        // before와 after 동시 사용 불가
        if (before != null && after != null) {
            throw BusinessException(ErrorCode.INVALID_INPUT_VALUE, "before와 after 파라미터는 동시에 사용할 수 없습니다")
        }

        val messages = when {
            // after: 놓친 메시지 조회 (오래된 순)
            after != null -> {
                val pageable = PageRequest.of(0, size, Sort.by(Sort.Direction.ASC, "createdAt"))
                chatMessageRepository.findByChatRoomIdAndIsDeletedFalseAndCreatedAtAfterOrderByCreatedAtAsc(
                    chatRoomId,
                    after,
                    pageable
                )
            }
            // before: 과거 메시지 조회 (최신순)
            before != null -> {
                val pageable = PageRequest.of(0, size, Sort.by(Sort.Direction.DESC, "createdAt"))
                chatMessageRepository.findByChatRoomIdAndIsDeletedFalseAndCreatedAtBeforeOrderByCreatedAtDesc(
                    chatRoomId,
                    before,
                    pageable
                )
            }
            // 둘 다 null: 최신 메시지 조회 (최신순)
            else -> {
                val pageable = PageRequest.of(0, size, Sort.by(Sort.Direction.DESC, "createdAt"))
                chatMessageRepository.findByChatRoomIdAndIsDeletedFalseOrderByCreatedAtDesc(
                    chatRoomId,
                    pageable
                )
            }
        }

        messages.map { message ->
            val replyMessage = message.replyToMessageId?.let { replyId ->
                chatMessageRepository.findById(replyId).orElse(null)
            }
            ChatMessageResponse.from(message, replyMessage)
        }
    }

    /**
     * 채팅방에서 메시지 검색
     *
     * 예전에는 이 자리를 코루틴이 맡았다. 밑에 있는 저장소가 전부 블로킹이라
     * 코루틴이 사 주는 것이 없었고, 톰캣 워커를 잡아 둔 채 IO 스레드를 하나 더
     * 쓰고 있었다. 지금은 그냥 부른다.
     *
     * @param chatRoomId 채팅방 ID
     * @param keyword 검색어
     * @param page 페이지 번호
     * @param size 페이지 크기
     * @param memberId 요청한 회원 ID (권한 확인용)
     * @return 검색 결과
     */
    fun searchMessages(
        chatRoomId: Long,
        keyword: String,
        page: Int,
        size: Int,
        memberId: Long
    ): List<ChatMessageResponse> = run {
        // 권한 확인
        validateChatRoomAccess(chatRoomId, memberId)
        val pageable = PageRequest.of(page, size, Sort.by(Sort.Direction.DESC, "createdAt"))

        chatMessageRepository
            .findByChatRoomIdAndIsDeletedFalseAndContentContainingOrderByCreatedAtDesc(
                chatRoomId,
                keyword,
                pageable
            )
            .map { message ->
                val replyMessage = message.replyToMessageId?.let { replyId ->
                    chatMessageRepository.findById(replyId).orElse(null)
                }
                ChatMessageResponse.from(message, replyMessage)
            }
    }

    /**
     * 안읽은 메시지 개수 조회
     *
     * @param chatRoomId 채팅방 ID
     * @param lastReadAt 마지막으로 읽은 시간
     * @return 안읽은 메시지 개수
     */
    fun getUnreadCount(chatRoomId: Long, lastReadAt: Instant?): Long =
        run {
            if (lastReadAt == null) {
                return 0L
            }

            chatMessageRepository.countByChatRoomIdAndIsDeletedFalseAndCreatedAtAfter(chatRoomId, lastReadAt)
        }

    /**
     * 특정 시간 이후의 메시지 조회 (재연결 시 놓친 메시지 복구용)
     *
     * WebSocket 연결이 끊겼다가 재연결될 때,
     * 마지막으로 받은 메시지 이후의 모든 메시지를 조회
     *
     * @param chatRoomId 채팅방 ID
     * @param after 이 시간 이후의 메시지 조회
     * @param limit 최대 개수 (기본 100)
     * @param memberId 요청한 회원 ID (권한 확인용)
     * @return 놓친 메시지 목록 (오래된 순)
     */
    fun getMessagesAfter(
        chatRoomId: Long,
        after: Instant,
        limit: Int = 100,
        memberId: Long
    ): List<ChatMessageResponse> = run {
        // 권한 확인
        validateChatRoomAccess(chatRoomId, memberId)
        val pageable = PageRequest.of(0, limit, Sort.by(Sort.Direction.ASC, "createdAt"))

        chatMessageRepository
            .findByChatRoomIdAndIsDeletedFalseAndCreatedAtAfterOrderByCreatedAtAsc(
                chatRoomId,
                after,
                pageable
            )
            .map { message ->
                val replyMessage = message.replyToMessageId?.let { replyId ->
                    chatMessageRepository.findById(replyId).orElse(null)
                }
                ChatMessageResponse.from(message, replyMessage)
            }
    }

    /**
     * 특정 메시지 주변 조회 (답장 점프, 검색 결과 점프용)
     *
     * 특정 메시지를 중심으로 앞뒤 메시지를 함께 조회
     * - before개: 해당 메시지 이전 메시지들 (과거)
     * - target: 해당 메시지
     * - after개: 해당 메시지 이후 메시지들 (미래)
     *
     * @param chatRoomId 채팅방 ID
     * @param messageId 중심이 될 메시지 ID
     * @param before 이전 메시지 개수 (기본 20)
     * @param after 이후 메시지 개수 (기본 20)
     * @param memberId 요청한 회원 ID (권한 확인용)
     * @return 중심 메시지 + 주변 메시지 목록 (시간순 정렬)
     */
    fun getMessagesAround(
        chatRoomId: Long,
        messageId: String,
        before: Int = 20,
        after: Int = 20,
        memberId: Long
    ): List<ChatMessageResponse> = run {
        // 권한 확인
        validateChatRoomAccess(chatRoomId, memberId)

        // 중심 메시지 조회
        val targetMessage = chatMessageRepository.findById(messageId)
            .orElseThrow { BusinessException(ErrorCode.RESOURCE_NOT_FOUND, "메시지를 찾을 수 없습니다") }

        // 채팅방 확인
        if (targetMessage.chatRoomId != chatRoomId) {
            throw BusinessException(ErrorCode.INVALID_INPUT_VALUE, "해당 채팅방의 메시지가 아닙니다")
        }

        val targetTime = targetMessage.createdAt
            ?: throw BusinessException(ErrorCode.INTERNAL_SERVER_ERROR, "메시지 생성 시간이 없습니다")

        // 1. 이전 메시지 조회 (과거 → 현재 방향으로 before개)
        val beforeMessages = if (before > 0) {
            val pageable = PageRequest.of(0, before, Sort.by(Sort.Direction.DESC, "createdAt"))
            chatMessageRepository.findByChatRoomIdAndIsDeletedFalseAndCreatedAtBeforeOrderByCreatedAtDesc(
                chatRoomId,
                targetTime,
                pageable
            ).reversed() // 시간순으로 뒤집기
        } else {
            emptyList()
        }

        // 2. 이후 메시지 조회 (현재 → 미래 방향으로 after개)
        val afterMessages = if (after > 0) {
            val pageable = PageRequest.of(0, after, Sort.by(Sort.Direction.ASC, "createdAt"))
            chatMessageRepository.findByChatRoomIdAndIsDeletedFalseAndCreatedAtAfterOrderByCreatedAtAsc(
                chatRoomId,
                targetTime,
                pageable
            )
        } else {
            emptyList()
        }

        // 3. 모든 메시지 합치기 (이전 메시지 + 중심 메시지 + 이후 메시지)
        val allMessages = beforeMessages + listOf(targetMessage) + afterMessages

        // 4. DTO 변환 (답장 정보 포함)
        allMessages.map { message ->
            val replyMessage = message.replyToMessageId?.let { replyId ->
                chatMessageRepository.findById(replyId).orElse(null)
            }
            ChatMessageResponse.from(message, replyMessage)
        }
    }

    /**
     * 메시지 삭제 (Soft Delete) - 현업 표준 버전
     *
     * 예전에는 이 자리를 코루틴이 맡았다. 밑에 있는 저장소가 전부 블로킹이라
     * 코루틴이 사 주는 것이 없었고, 톰캣 워커를 잡아 둔 채 IO 스레드를 하나 더
     * 쓰고 있었다. 지금은 그냥 부른다.
     *
     * 개선 사항:
     * - 삭제 후 Redis Pub/Sub 발행 (실시간 동기화)
     * - 답장 정보 포함
     * - 에러 처리 강화
     *
     * @param chatRoomId 채팅방 ID
     * @param messageId 메시지 ID
     * @param memberId 요청한 회원 ID (권한 확인용)
     */
    fun deleteMessage(chatRoomId: Long, messageId: String, memberId: Long) =
        run {
            try {
                val message = chatMessageRepository.findById(messageId)
                    .orElseThrow { BusinessException(ErrorCode.RESOURCE_NOT_FOUND, "메시지를 찾을 수 없습니다") }

                // 채팅방 확인
                if (message.chatRoomId != chatRoomId) {
                    throw BusinessException(ErrorCode.INVALID_INPUT_VALUE, "해당 채팅방의 메시지가 아닙니다")
                }

                // 본인 메시지만 삭제 가능
                if (message.senderId != memberId) {
                    throw BusinessException(ErrorCode.FORBIDDEN, "본인의 메시지만 삭제할 수 있습니다")
                }

                // 이미 삭제된 메시지
                if (message.isDeleted) {
                    throw BusinessException(ErrorCode.INVALID_INPUT_VALUE, "이미 삭제된 메시지입니다")
                }

                // Soft Delete
                message.delete()
                val savedMessage = chatMessageRepository.save(message)

                logger.info(
                    "메시지 삭제 완료: messageId={}, chatRoomId={}, memberId={}",
                    messageId,
                    chatRoomId,
                    memberId
                )

                // 답장 메시지 조회
                val replyMessage = savedMessage.replyToMessageId?.let { replyId ->
                    chatMessageRepository.findById(replyId).orElse(null)
                }

                // 수신자 ID 계산 (실시간 동기화용)
                val chatRoom = chatRoomRepository.findById(chatRoomId)
                    .orElseThrow { BusinessException(ErrorCode.RESOURCE_NOT_FOUND, "채팅방을 찾을 수 없습니다") }

                val receiverId = if (memberId == chatRoom.buyer.memberId) {
                    chatRoom.seller.memberId!!
                } else {
                    chatRoom.buyer.memberId!!
                }

                // Redis Pub/Sub 발행 (실시간 동기화)
                val response = ChatMessageResponse.from(savedMessage, replyMessage).toBuilder().receiverId(receiverId).build()
                redisPubSubPublisher.publish(response)

                logger.debug("메시지 삭제 이벤트 발행: messageId={}", messageId)

            } catch (e: BusinessException) {
                throw e
            } catch (e: Exception) {
                logger.error(
                    "메시지 삭제 실패: messageId={}, chatRoomId={}, memberId={}, error={}",
                    messageId,
                    chatRoomId,
                    memberId,
                    e.message,
                    e
                )
                throw BusinessException(ErrorCode.INTERNAL_SERVER_ERROR, "메시지 삭제 중 오류가 발생했습니다")
            }
        }

    /**
     * 메시지 수정 - 현업 표준 버전
     *
     * 예전에는 이 자리를 코루틴이 맡았다. 밑에 있는 저장소가 전부 블로킹이라
     * 코루틴이 사 주는 것이 없었고, 톰캣 워커를 잡아 둔 채 IO 스레드를 하나 더
     * 쓰고 있었다. 지금은 그냥 부른다.
     *
     * 개선 사항:
     * - 첫 수정 시 원본 내용 저장 (originalContent)
     * - 수정 시간 기록 (updatedAt)
     * - 수정 플래그 설정 (isEdited)
     * - Redis Pub/Sub 발행 (실시간 동기화)
     * - 답장 정보 포함
     * - 에러 처리 강화
     *
     * @param chatRoomId 채팅방 ID
     * @param messageId 메시지 ID
     * @param memberId 요청한 회원 ID (권한 확인용)
     * @param newContent 수정할 내용
     * @return 수정된 메시지 DTO
     */
    fun updateMessage(
        chatRoomId: Long,
        messageId: String,
        memberId: Long,
        newContent: String
    ): ChatMessageResponse = run {
        try {
            // 공백 검증
            if (newContent.isBlank()) {
                throw BusinessException(ErrorCode.INVALID_INPUT_VALUE, "메시지 내용은 비어 있을 수 없습니다")
            }

            val message = chatMessageRepository.findById(messageId)
                .orElseThrow { BusinessException(ErrorCode.RESOURCE_NOT_FOUND, "메시지를 찾을 수 없습니다") }

            // 채팅방 확인
            if (message.chatRoomId != chatRoomId) {
                throw BusinessException(ErrorCode.INVALID_INPUT_VALUE, "해당 채팅방의 메시지가 아닙니다")
            }

            // 본인 메시지만 수정 가능
            if (message.senderId != memberId) {
                throw BusinessException(ErrorCode.FORBIDDEN, "본인의 메시지만 수정할 수 있습니다")
            }

            // 텍스트 메시지만 수정 가능
            if (message.type != MessageType.TEXT) {
                throw BusinessException(ErrorCode.INVALID_INPUT_VALUE, "텍스트 메시지만 수정할 수 있습니다")
            }

            // 삭제된 메시지는 수정 불가
            if (message.isDeleted) {
                throw BusinessException(ErrorCode.INVALID_INPUT_VALUE, "삭제된 메시지는 수정할 수 없습니다")
            }

            // 내용이 동일하면 수정하지 않음
            if (message.content == newContent) {
                throw BusinessException(ErrorCode.INVALID_INPUT_VALUE, "수정할 내용이 기존 내용과 동일합니다")
            }

            // 수정 (MongoDB data class는 불변이므로 새 객체 생성)
            val updatedMessage = message.copy(
                content = newContent,
                updatedAt = Instant.now(),
                isEdited = true,
                // 첫 수정 시에만 원본 저장 (이후 수정은 최초 원본 유지)
                originalContent = message.originalContent ?: message.content
            )

            val savedMessage = chatMessageRepository.save(updatedMessage)

            logger.info(
                "메시지 수정 완료: messageId={}, chatRoomId={}, memberId={}, isFirstEdit={}",
                messageId,
                chatRoomId,
                memberId,
                message.originalContent == null
            )

            // 답장 메시지 조회
            val replyMessage = savedMessage.replyToMessageId?.let { replyId ->
                chatMessageRepository.findById(replyId).orElse(null)
            }

            // 수신자 ID 계산 (실시간 동기화용)
            val chatRoom = chatRoomRepository.findById(chatRoomId)
                .orElseThrow { BusinessException(ErrorCode.RESOURCE_NOT_FOUND, "채팅방을 찾을 수 없습니다") }

            val receiverId = if (memberId == chatRoom.buyer.memberId) {
                chatRoom.seller.memberId!!
            } else {
                chatRoom.buyer.memberId!!
            }

            // Redis Pub/Sub 발행 (실시간 동기화)
            val response = ChatMessageResponse.from(savedMessage, replyMessage).toBuilder().receiverId(receiverId).build()
            redisPubSubPublisher.publish(response)

            logger.debug("메시지 수정 이벤트 발행: messageId={}", messageId)

            response

        } catch (e: BusinessException) {
            throw e
        } catch (e: Exception) {
            logger.error(
                "메시지 수정 실패: messageId={}, chatRoomId={}, memberId={}, error={}",
                messageId,
                chatRoomId,
                memberId,
                e.message,
                e
            )
            throw BusinessException(ErrorCode.INTERNAL_SERVER_ERROR, "메시지 수정 중 오류가 발생했습니다")
        }
    }
}